### PR TITLE
scenario editing mode improvements and new unique

### PR DESCRIPTION
**Pull request by HadeanLake**
* Fixed Russian nation unique - it is civ-wide bonus

* Fixed crash in scenario editing mode when selecting to produce gold/science/nothing per turn

* genericized "+[]% Production when building [] in this city"

* genericized "All newly-trained [] units in this city receive the [] promotion"

* scenario editing mode improvements:

Resource requirements when buying buildings and units are ignored

Limit for selling buildings is ignored

Workers improve tiles in one turn

Units will not be disbanded if civilization has negative gold

### CODE REVIEW

**Comment by yairm210** [comment]
If the parameters are come delimited they didn't be translated correctly

**Comment by HadeanLake** [comment]
`"uniques": ["All newly-trained [Melee] units in this city receive the [Drill I] promotion", "All newly-trained [Mounted] units in this city receive the [Drill I] promotion", "All newly-trained [Armor] units in this city receive the [Drill I] promotion"],`
that will be a very long unique then

**Comment by yairm210** [comment]
"All newly-trained relevant units" perhaps, as in 'units that can receive'?

File: core/src/com/unciv/logic/civilization/CivilizationInfo.kt [hunk 1]
```diff
@@ -436,7 +436,7 @@ class CivilizationInfo {
         policies.endTurn(nextTurnStats.culture.toInt())
 
         // disband units until there are none left OR the gold values are normal
-        if (!isBarbarian() && gold < -100 && nextTurnStats.gold.toInt() < 0) {
+        if (!isBarbarian() && !gameInfo.gameParameters.godMode && gold < -100 && nextTurnStats.gold.toInt() < 0) {
```
Comment (yairm210) [inline core/src/com/unciv/logic/civilization/CivilizationInfo.kt:439]: These extra units would just disband anyway on the first turn of the player, not sure what we gain here

File: core/src/com/unciv/logic/city/CityStats.kt [hunk 2]
```diff
@@ -309,6 +311,20 @@ class CityStats {
                 stats.production += 15
         }
 
+        for (unique in cityInfo.cityConstructions.builtBuildingUniqueMap.getUniques("+[]% production when building [] in this city")) {
+            val filter = unique.params[1]
+            if (currentConstruction.name == filter
+                    || (filter == "land units" && currentConstruction is BaseUnit && currentConstruction.unitType.isLandUnit())
+                    || (filter == "naval units" && currentConstruction is BaseUnit && currentConstruction.unitType.isWaterUnit())
```
Comment (yairm210) [inline core/src/com/unciv/logic/city/CityStats.kt:318]: "naval units" should appear in template.properties for translations if it's going to appear as a placeholder parameter

File: core/src/com/unciv/logic/city/CityStats.kt [hunk 3]
```diff
@@ -309,6 +311,20 @@ class CityStats {
...
+                    || (filter == "mounted units" && currentConstruction is BaseUnit && currentConstruction.unitType == UnitType.Mounted)
```
Comment (yairm210) [inline core/src/com/unciv/logic/city/CityStats.kt:319]: Same with this - unless we plan on making this a separate unique that accepts a unitType
"+[]% production when building [] units in this city"
This connects to another big dilemma I have about unit filtering as a whole

File: core/src/com/unciv/models/ruleset/unit/BaseUnit.kt [hunk 4]
```diff
@@ -164,6 +164,17 @@ class BaseUnit : INamed, IConstruction {
             XP += unique.params[0].toInt()
         unit.promotions.XP = XP
 
+        for (unique in construction.cityInfo.cityConstructions.builtBuildingUniqueMap.getUniques("All newly-trained [] units in this city receive the [] promotion")) {
+            val filter = unique.params[0]
+            val promotion = unique.params[1]
+            if (unit.name == filter
+                    || (filter == "relevant" && civInfo.gameInfo.ruleSet.unitPromotions.values.any { unit.type.toString() in it.unitTypes && it.name == promotion })
+                    || (unit.type.toString() == filter)
```
Comment (yairm210) [inline core/src/com/unciv/models/ruleset/unit/BaseUnit.kt:172]: unit.type.name
and remove the unnecessary parentheses :)

File: core/src/com/unciv/logic/civilization/CivilizationInfo.kt (hunk 1 above)
Comment (HadeanLake) [reply core/src/com/unciv/logic/civilization/CivilizationInfo.kt:439]: Scenario creators are not annoyed while testing something. 
But yes will disband if scenario creator did not check gold before releasing scenario.

Comment (HadeanLake) [reply core/src/com/unciv/logic/civilization/CivilizationInfo.kt:439]: well maybe having units deleted will help them not to forget to check and edit in save gold amount

**Pull request merged**